Add utilities

bjorn-stevens opened the pull request:
Added a wrapper for one of the examples and a utilities module where a small set of useful things can be collected.  Update documentation in examples accordingly

bjorn-stevens committed:
added tolerances for integration

bjorn-stevens committed:
add wraper for moist adiabat

added a utilities function for things that might often be
useful.  Started out by adding the example where we integrate
the moist adiabt numerically and include the ice phase

also updated the examples.ipynb to reflect the new changes
and correct an error associated with the use of the new
way of integrating the moist adiabat that was merged previously

bjorn-stevens commented:
These are fairly minimal changes.  If this is okay and we merge I would then suggest to release a new version and we update the packages that use some of this.  As part of the review, and new version, maybe look at the non-hydrostatic inegration routine.  I'm not so happy with the naming of that and probably we need documentation.  Let's see who is the first to give it some thought.

bjorn-stevens committed:
modified tests inlight of name change

bjorn-stevens committed:
fixup naming of brunt vaisala frequency and tests

lkluft commented on moist_thermodynamics/utilities.py:
```diff
@@ -0,0 +1,49 @@
+            "Tice": {"cx": constants.ci, "lx": mt.sublimation_enthalpy},
+        }
+        es = mt.make_es_mxd(es_liq=svp.liq_analytic, es_ice=svp.ice_analytic)
+        for y in X.values():
+            y["T"], Px = mt.moist_adiabat(
```
I have to admit that I am a bit lost in this function in general. But it seems that here, you are looping over the constant for liquid and ice, but only store references to the latter. Is this in purpose or do you actually want to store both results?

lkluft replied:
The `y["T"]` notation will add a new `"T"` entry to the existing dictionaries. Those are then reused later.

lkluft commented on moist_thermodynamics/utilities.py:
```diff
@@ -0,0 +1,49 @@
+        TK = np.ones(len(Px)) * T0
+        TK[X["Tliq"]["T"] > T0] = X["Tliq"]["T"][X["Tliq"]["T"] > T0]
+        TK[X["Tice"]["T"] < T0] = X["Tice"]["T"][X["Tice"]["T"] < T0]
+    else:
+        TK = np.vectorize(mt.invert_for_temperature)(thx, thx(Tx, P[0], qx), P, qx)
```
The second implementation of the function does compute the temperature profile that would have been needed to get an isentropic profile for a given $p$ and $q$?

These seem like two very different approaches. Do we prefer one over? If so, can we remove the other?

lkluft reviewed (approved):
I left some inline comment for the `get_mosit_adiabat` wrapper.

As this is just an addition and doesn't affect any of the "library" functions, I would be fine to just merge a release a new version to use in the other package.

In general, we should really move to examples.docs that are build automatically. I think it is nice to provide hands-on examples, but we **have** to generate them as part of the CI, otherwise they will break all the time (like ICON examples in various places).

lkluft commented:
The "approval" might have been a bit over-enthusiastic. I would like to at least understand what looping over the constants in `get_moist_adiabat` does 😉

bjorn-stevens committed:
fixup of functions

the default values of pressure altitude were incorrectly specified
as scalars, this has been fixed

the moist adiabat was restructured so that output need not assume
a constant pressure grid, simplifying argument lists.

bjorn-stevens committed:
fix tests and ruff

bjorn-stevens committed:
attempt fix of test case;5D;5D

bjorn-stevens committed:
ruff fix

bjorn-stevens committed:
ruff fix

bjorn-stevens committed:
test fix

bjorn-stevens committed:
test fix

bjorn-stevens commented:
I had a problem with the test environment, which needed me to pass a tuple of values.  This works now.  So I will merge.
